ppcs/ppcTODO-attributes-v2.md

```diff
@@ -0,0 +1,216 @@
+
+This might suggest that actually the target should be specified as two - or maybe even three - parameters, some of which would be zero / NULL depending on circumstance.
+
+```c
+    void (*apply)(pTHX_
+        SV *target, GV *targetname, /* for package targets, or NULL/NULL */
+        PADOFFSET targetix,         /* for lexical targets, or zero */
+        SV *attrvalue);
+```
+
+But perhaps at that point, it makes more sense to have two different callbacks, one for GV-named targets and one for lexicals?
+
+```c
+struct PerlAttributeDefinition
+{
+    ...
+    void (*apply_pkg)(pTHX_ SV *target, GV *targetname, SV *attrvalue);
+    void (*apply_lex)(pTHX_ PADOFFSET targetix,         SV *attrvalue);
+};
+```
+
+Or perhaps one function that takes a distinguishing enum and union type to convey the information:
+
+```c
+enum PerlAttributeTargetKind {
+    ATTRTARGET_PKGSCOPED,
+    ATTRTARGET_LEXICAL,
+};
+
+union PerlAttributeTarget {
+    struct { SV *sv; GV *namegv; } pkgscoped;
+    struct { PADOFFSET padix;    } lexical;
+};
+
+struct PerlAttributeDefinition
+{
+    ...
+    void (*apply)(pTHX_ 
+        enum PerlAttributeTargetKind kind, union PerlAttributeTarget target,
+        SV *attrvalue);
+};
+```
+
+### Split the Pad into Scope + Scratchpad
+
+While not unique to this proposal, the ongoing increase in use of lexical imports in various parts of perl means that pads in typical programs - both at the file and subroutine level - are getting wider, with more named items in there. Because currently the pads are shared with true per-call lexicals and temporaries, this means that any recursive functions consume more space in unnecessary elements. Every recursive depth of function call requires the entire width of the pad to be cloned for each level. The more "static" elements in the pad, the more wasted space because those elements are not going to vary with depth.
+
+It would be a nontrivial undertaking, but it may become useful in terms of memory (and CPU cycles) savings to consider splitting the pad into two separate entities. A "scope" pad would exist just once per file or subroutine, and contain lexically-imported named elements such as imported functions or attribute definitions. Separately a "scratchpad" pad would then only contain the lexical variables and temporary values that are needed once for every depth of recursion of the function. By splitting the pad in such a way, it reduces the amount of work that `pp_entersub` has to perform when recursing into functions that have wide pads because now only the true variables-per-call need to be cloned; the static scope would exist just once for all depths.
+
+### Perl-Level API
+
+The mechanisms described in this document exist entirely at the C level, intended for XS modules to make use of. It should be possible to provide wrappings of most of the relevant parts, at least for simple cases, for use directly by Perl authors without writing C code.
```

Comment:
I'd suggest to define reusable (from core as well from 3rd party modules) syntax to declare additional symbols (taken from my "independent symbol spaces"). For attributes:
```
declare attribute Foo as
    on CODE   => { ... do something ... }
    on SCALAR => { ... do something ... }
    on ARRAY | HASH => { ... do something ... }
  ;
```